src/main/scala/org/apache/spark/mllib/optimization/tfocs/ProxCapableFunction.scala

```diff
@@ -20,36 +20,116 @@ package org.apache.spark.mllib.optimization.tfocs
 import org.apache.spark.mllib.linalg.{ DenseVector, Vector, Vectors }
 
 /**
- * Trait for prox-capable functions.
+ * A trait for prox capable functions which support efficient proximity minimization, as expressed
+ * by the proximity operator:
+ *   x = prox_h(z, t) = argmin_x(h(x) + 0.5 * ||x - z||_2^2 / t)
+ *
+ * Both the minimizing x value and the function value h(x) may be computed, depending on the
+ * mode specified.
+ *
+ * @tparam X A type representing a vector on which to evaluate the function.
  */
 trait ProxCapableFunction[X] {
-  /**
-   * Evaluates this function at x with smoothing parameter t.
-   */
-  def apply(x: X, t: Double, mode: Mode): Value[Double, X]
 
   /**
-   * Evaluates this function at x.
+   * Evaluate the proximity operator prox_h at z with parameter t, returning both x and h(x)
+   * depending on the mode specified.
+   *
+   * @param z The vector on which to evaluate the proximity operator.
+   * @param t The proximity parameter.
+   * @param mode The computation mode. If mode.f is true, h(x) is returned. If mode.g is true, x
+   *        is returned.
+   *
+   * @return A Value containing x, the vector minimizing the proximity function prox_h, and/or h(x),
+   *         the function value at x. The exact list of values computed and returned depends on the
+   *         attributes of the supplied 'mode' parameter. The returned Value contains h(x) in its
+   *         'f' attribute, while x is contained in the 'g' attribute.
```

Comment:
`Value` is to store `f` (function value) and `g` (gradient value). Is `x` the gradient here? If not, we should define a different container for the result type.